src/v/cluster/feature_manager.cc

```diff
@@ -288,4 +288,63 @@ ss::future<> feature_manager::do_maybe_update_active_version() {
     vlog(clusterlog.info, "Updated cluster version to {}", max_version);
 }
 
+ss::future<std::error_code>
+feature_manager::write_action(cluster::feature_update_action action) {
+    // TODO: validate earlier that the feature name is known to us
+
+    // Validate that teh feature is in a state compatible with the
+    // requested transition.
+    bool valid = true;
+    auto state = _feature_table.local().get_state(action.feature_name);
+    switch (action.action) {
+    case cluster::feature_update_action::action_t::complete_preparing:
+        // Look up feature by name
+        if (state.get_state() != feature_state::state::preparing) {
+            // Drop this silently, we presume that this is some kind of
+            // race and the thing we thought was preparing is either
+            // now active or administratively deactivated.
+            valid = false;
+        }
+
+        break;
+    case cluster::feature_update_action::action_t::administrative_activate:
```

Comment:
Doing it here is a pre-check rather than a strict correctness thing: it avoids the noise of writing something to disk that may later be ignored/rejected by the backend.

The primary check is in `feature_table::apply_action`, where we drop updates if they try to transition the feature to a state that is not valid from its current state.